src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLStyleVisitor.java

```diff
@@ -0,0 +1,301 @@
+/* (c) 2024 Open Source Geospatial Foundation - all rights reserved
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.mapml;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.geotools.api.filter.Filter;
+import org.geotools.api.filter.expression.Expression;
+import org.geotools.api.style.FeatureTypeStyle;
+import org.geotools.api.style.Fill;
+import org.geotools.api.style.Graphic;
+import org.geotools.api.style.GraphicalSymbol;
+import org.geotools.api.style.LineSymbolizer;
+import org.geotools.api.style.Mark;
+import org.geotools.api.style.PointSymbolizer;
+import org.geotools.api.style.PolygonSymbolizer;
+import org.geotools.api.style.RasterSymbolizer;
+import org.geotools.api.style.Rule;
+import org.geotools.api.style.Stroke;
+import org.geotools.api.style.Symbol;
+import org.geotools.api.style.Symbolizer;
+import org.geotools.api.style.TextSymbolizer;
+import org.geotools.filter.visitor.IsStaticExpressionVisitor;
+import org.geotools.styling.AbstractStyleVisitor;
+import org.geotools.util.logging.Logging;
+
+/** MapML Style Visitor to convert SLD to MapML CSS Styles */
+public class MapMLStyleVisitor extends AbstractStyleVisitor {
+    static final Logger LOGGER = Logging.getLogger(MapMLStyleVisitor.class);
+    public static final String OPACITY = "opacity";
+
+    /** Default radius for point symbolizers, see @link{Graphic#getSize()} */
+    private static final Double DEFAULT_RADIUS = 8.0;
+
+    public static final String STROKE_OPACITY = "stroke-opacity";
+    public static final String STROKE_DASHARRAY = "stroke-dasharray";
+    public static final String STROKE_LINECAP = "stroke-linecap";
+    public static final String STROKE_WIDTH = "stroke-width";
+    public static final String STROKE = "stroke";
+    public static final String FILL = "fill";
+    public static final String RADIUS = "r";
+    public static final String RULE_SYMBOLIZER_DELIMITER = ":";
+    public static final String STROKE_DASHOFFSET = "stroke-dashoffset";
+    public static final String FILL_OPACITY = "fill-opacity";
+
+    Map<String, MapMLStyle> styles = new HashMap<>();
+    AtomicInteger ruleCounter = new AtomicInteger(0);
+    AtomicInteger symbolizerCounter = new AtomicInteger(0);
+    AtomicBoolean isElseFilter = new AtomicBoolean(false);
```

Comment:
I don't think you need anything atomic here, it's not a concurrent class, and these are fields, so they can be updated without issue... e.g., it should work with simple int and boolean variables?

Reply:
Switched to simple int and boolean